feat: RC1 what's new update (#4115)
JetBrains/kotlin-web-site: docs/topics/whatsnew-eap.md

* [Enable the Send usage statistics option](https://www.jetbrains.com/help/idea/settings-usage-statistics.html) to
  allow JetBrains to collect anonymous data about K2 usage.

## Kotlin/JVM

This version brings the following changes:

> * [Generate lambda functions using invokedynamic](#generate-lambda-functions-using-invokedynamic)

Reviewer: ```suggestion
* [Generate lambda functions using `invokedynamic`](#generate-lambda-functions-using-invokedynamic)
```

> * [The kotlinx-metadata-jvm library is now Stable](#the-kotlinx-metadata-jvm-library-is-now-stable)

Reviewer: ```suggestion
* [The `kotlinx-metadata-jvm` library is now Stable](#the-kotlinx-metadata-jvm-library-is-now-stable)
```

### Generate lambda functions using invokedynamic

Kotlin %kotlinEapVersion% introduces a new default method for generating lambda functions using `invokedynamic`.
This change reduces the binary sizes of applications compared to the traditional anonymous class generation.

Since the first version, Kotlin has generated lambdas as anonymous classes. However, starting from [Kotlin 1.5](whatsnew15.md#lambdas-via-invokedynamic),
the option for `invokedynamic` generation was available by using the `-Xlambdas=indy` compiler flag. In Kotlin %kotlinEapVersion%,
`invokedynamic` has become the default method for lambda generation. This method produces lighter binaries and aligns
Kotlin with JVM optimizations, ensuring applications benefit from ongoing and future improvements in JVM performance.

Currently, it has three limitations compared to ordinary lambda compilation:

* A lambda compiled into `invokedynamic` is not serializable.

> * Calling `toString()` on such a lambda produces a less readable string representation.

Reviewer 2: Sure, we can put it as the last point, but I don't think it can be clarified in a simple way without including a long example which would make it far less readable. 

I think we should include it after it leaves RC in the docs, but in the what's new section, it would feel a bit too much to me.

  support lambdas generated by `invokedynamic`.

To retain the legacy behavior of generating lambda functions, you can either:

* Annotate specific lambdas with `@JvmSerializableLambda`.
* Use the compiler argument `-Xlambdas=class` to generate all lambdas in a module using the legacy method.

### The kotlinx-metadata-jvm library is now Stable

In %kotlinEapVersion%, the `kotlinx-metadata-jvm` library became [Stable](components-stability.md#stability-levels-explained).
Since the library changed to the `kotlin` package and coordinates, now you can find it as `kotlin-metadata-jvm` (without the "x").

Before, the `kotlinx-metadata-jvm` library had its own publishing scheme and version. Now, we will build and publish the
`kotlin-metadata-jvm` updates as part of the Kotlin release cycle, with the same backward compatibility guarantees as the
Kotlin standard library.

The `kotlin-metadata-jvm` library provides an API to read and modify metadata of binary files generated by the Kotlin/JVM
compiler.

## Kotlin/Native

This version brings the following changes:

* [Resolving conflicts with Objective-C methods](#resolving-conflicts-with-objective-c-methods)

> * [Changed log level for compiler arguments in Kotlin/Native](#changed-log-level-for-compiler-arguments)

Author: I checked, and we don't have consistency in these header lists throughout different WNs, as well as this one. I'll revisit it in RC2
For this section in particular, I'd like to keep the sense(usefulness) of the feature in the header, for example, "Conflict resolution with Objective-C methods", but change the second header to look similar

* [Improved Gradle dependency handling for CInteropProcess in Kotlin/Native](#improved-gradle-dependency-handling-for-cinteropprocess-in-kotlin-native)
* [Visibility changes in Gradle](#visibility-changes-in-gradle)
* [New directory for Kotlin data in Gradle projects](#new-directory-for-kotlin-data-in-gradle-projects)
* [Kotlin/Native compiler downloaded when needed](#kotlin-native-compiler-downloaded-when-needed)
* [Deprecating old ways of defining compiler options](#deprecating-old-ways-of-defining-compiler-options)

### New Gradle DSL for compiler options in multiplatform projects

> This feature is [Experimental](components-stability.md#stability-levels-explained). It may be dropped or changed at any time.
> Use it only for evaluation purposes. We would appreciate your feedback on it in [YouTrack](https://kotl.in/issue).
>
{type="warning"}

Prior to Kotlin %kotlinEapVersion%, configuring compiler options in a multiplatform project with Gradle was only possible
at a low level, such as per task, compilation, or source set. To make it easier to configure compiler options more generally
in your projects, Kotlin %kotlinEapVersion% comes with a new Gradle DSL.

> With this new DSL, you can configure compiler options at the extension level for all the targets and shared source sets
> like `commonMain` and at a target level for a specific target:

Reviewer: ```suggestion
With this new DSL, you can configure compiler options at the extension level for all the targets and shared source sets,
such as `commonMain`, as well as at a target level for a specific target:
```

> ```kotlin

Reviewer 3: I feel that the example comes between two parts of the conceptual explanation (and that leads to "extension level" being explained later than it is mentioned). More natural sequence would be "... was only possible at low level → project configuration now has three layers → here's an example → here's nuance about conventions".

Author: I think this sequence would work for the actual documentation. Here, in What's new, we need to announce the news, the new DSL. The whole section starting with "3 layers" is additional information to explain how the overall picture looks like now, but it can be removed without losing the point of the announcement

Reviewer 3: My main point is about the "extension level"; if the intended audience doesn't need the term introduced here, then it's all good.

    compilerOptions {
        // Extension-level common compiler options that are used as defaults
        // for all targets and shared source sets
        allWarningsAsErrors.set(true)
    }
    jvm {
        compilerOptions {
            // Target-level JVM compiler options that are used as defaults
            // for all compilations in this target
            noJdk.set(true)
        }
    }
}
```

The overall project configuration now has three layers. The highest is the extension level, then the target level,
and the lowest is the compilation unit (which is usually a compilation task):

![Kotlin compiler options levels](compiler-options-levels.svg){width=700}

The settings at a higher level are used as a convention (defaults) for a lower level:

* The extension compiler options values are the default for target compiler options, including shared source sets,
  like `commonMain`, `nativeMain`, and `commonTest`.
* Target compiler options values are used as the default for compilation unit (task) compiler options, for example,
  `compileKotlinJvm` and `compileTestKotlinJvm` tasks.

In turn, the configuration made at a lower level overrides related settings at a higher level:

> 

Reviewer: ```suggestion
```

When configuring your project, keep in mind that some old ways of setting up compiler options were [deprecated](#deprecating-old-ways-of-defining-compiler-options).

We encourage you to try the new DSL out in your multiplatform projects and leave feedback in [YouTrack](https://kotl.in/issue),
as we plan to make this DSL the recommended approach for configuring compiler options.

### New attribute to distinguish JVM and Android published libraries

Starting with Kotlin %kotlinEapVersion%, the [`org.gradle.jvm.environment`](https://docs.gradle.org/current/userguide/variant_attributes.html#sub:jvm_default_attributes)
Gradle attribute is published by default with all Kotlin variants.

The attribute helps distinguish JVM and Android variants of Kotlin Multiplatform libraries. It indicates that a certain
library variant is better suited for a certain JVM environment. The target environment could be "android", "stardard-jvm",
or "no-jvm".

Publishing this attribute should make consuming Kotlin Multiplatform libraries with JVM and Android targets more robust
from non-multiplatform clients as well, such as Java-only projects.

> If necessary, you can disable attribute publication. To do that, add the following Gradle option to your `gradle.properties` file:

Reviewer 3: ```suggestion
If necessary, you can disable publication of this attribute. To do that, add the following Gradle option to your `gradle.properties` file:
```

So as not to say "disable publication of all attributes."

## What to expect from upcoming Kotlin EAP releases

> Starting from **Kotlin 2.0.0-RC1**, you can start using the K2 compiler in production.

Reviewer: ```suggestion
Starting from **Kotlin 2.0.0-RC1**, you can use the K2 compiler in production.
```
